File: test/fixtures/ts-app-merge/merge.test.js
import assert from 'node:assert/strict';
import { test } from 'node:test';
import { mergeCarts } from './merge.js';

test('mergeCarts combines quantities for the same item id', () => {
  const cartA = [
    { id: 'sku-1', qty: 2 },
    { id: 'sku-2', qty: 1 },
  ];
  const cartB = [
    { id: 'sku-1', qty: 3 },
    { id: 'sku-3', qty: 1 },
  ];

  const merged = mergeCarts(cartA, cartB);
  const bySku1 = merged.filter((i) => i.id === 'sku-1');

  // This is the seeded bug: sku-1 currently appears as two separate lines.
  assert.equal(bySku1.length, 1);
  assert.equal(bySku1[0].qty, 5);
  assert.equal(merged.length, 3);
});

test('carts with no overlap are unaffected', () => {
  const merged = mergeCarts([{ id: 'sku-1', qty: 1 }], [{ id: 'sku-2', qty: 1 }]);
  assert.equal(merged.length, 2);
});

File: test/fixtures/go-app-merge/merge_test.go
package merge

import "testing"

func TestMergeCartsCombinesQuantitiesForSameItemID(t *testing.T) {
	cartA := []Item{{ID: "sku-1", Qty: 2}, {ID: "sku-2", Qty: 1}}
	cartB := []Item{{ID: "sku-1", Qty: 3}, {ID: "sku-3", Qty: 1}}

	merged := MergeCarts(cartA, cartB)

	var bySku1 []Item
	for _, item := range merged {
		if item.ID == "sku-1" {
			bySku1 = append(bySku1, item)
		}
	}

	// This is the seeded bug: sku-1 currently appears as two separate lines.
	if len(bySku1) != 1 {
		t.Fatalf("len(bySku1) = %d, want 1", len(bySku1))
	}
	if bySku1[0].Qty != 5 {
		t.Errorf("bySku1[0].Qty = %d, want 5", bySku1[0].Qty)
	}
	if len(merged) != 3 {
		t.Errorf("len(merged) = %d, want 3", len(merged))
	}
}

func TestCartsWithNoOverlapAreUnaffected(t *testing.T) {
	merged := MergeCarts([]Item{{ID: "sku-1", Qty: 1}}, []Item{{ID: "sku-2", Qty: 1}})
	if len(merged) != 2 {
		t.Errorf("len(merged) = %d, want 2", len(merged))
	}
}
